src/basic/fd-util.c

```diff
@@ -615,3 +615,169 @@ int fd_move_above_stdio(int fd) {
         (void) close(fd);
         return copy;
 }
+
+int rearrange_stdio(int original_input_fd, int original_output_fd, int original_error_fd) {
+        _cleanup_close_ int new_input_fd = -1, new_output_fd = -1, new_error_fd = -1;
+        int input_fd, output_fd, error_fd, r;
+
+        /* Sets up stdin, stdout, stderr with the three file descriptors passed in. If any of the descriptors is
+         * specified as -1 it will be connected with /dev/null instead. If any of the file descriptors is passed as
+         * itself (e.g. stdin as STDIN_FILENO) it is left unmodified, but the O_CLOEXEC bit is turned off should it be
+         * on.
+         *
+         * Note that if any of the passed file descriptors are > 2 they will be closed — both on success and on
+         * failure! Thus, callers should assume that when this function returns the input fds are invalidated.
+         *
+         * Note that when this function fails stdin/stdout/stderr might remain half set up!
+         *
+         * O_CLOEXEC is turned off for all three file descriptors (which is how it should be for
+         * stdin/stdout/stderr). */
+
+        if (original_input_fd == STDIN_FILENO)
+                input_fd = -1; /* Nothing to do, fd is already at the right place */
+        else {
```

Comment:
Hmm, there's essentially the same code repeated three times. I think it'd be much better to add `rearrange_one_fd` for the part in line 636–662.